File: src/lib/entity.ts
import { Direction } from "@/lib/direction";
import { SpriteSheetID } from "@/lib/sprite";

export type Entity = {
  spriteSheetID: SpriteSheetID | null;
  frame: number;
  frameRowOffset: number;
  frameIncrement: number;
  frameCount: number;
  frameDirection: Map<Direction, number>;
  x: number;
  y: number;
  offsetX: number;
  offsetY: number;
  direction: Direction | null;
  directions: Direction[];
  radius: number;
  velocity: number;
  lastCorner: Entity | null;
  scaredTime: number;
};

export function newEntity({
  spriteSheetID = null,
  frame = 0,
  frameRowOffset = 0,
  frameIncrement = 0,
  frameCount = 0,
  frameDirection = new Map(),
  x = 0,
  y = 0,
  offsetX = 0,
  offsetY = 0,
  direction = null,
  directions = [],
  radius = 0,
  velocity = 0,
  lastCorner = null,
  scaredTime = 0,
}: Partial<Entity>): Entity {
  return {
    spriteSheetID,
    frame,
    frameRowOffset,
    frameIncrement,
    frameCount,
    frameDirection,
    x,
    y,
    offsetX,
    offsetY,
    direction,
    directions,
    radius,
    velocity,
    lastCorner,
    scaredTime,
  };
}

File: internal/entity.go
package internal

import "github.com/hajimehoshi/ebiten/v2"

type Direction string

const (
	DirectionUp    Direction = "UP"
	DirectionDown  Direction = "DOWN"
	DirectionLeft  Direction = "LEFT"
	DirectionRight Direction = "RIGHT"
)

type Axis string

const (
	AxisVertical   Axis = "VERTICAL"
	AxisHorizontal Axis = "HORIZONTAL"
)

func (d Direction) Axis() Axis {
	switch d {
	case DirectionUp, DirectionDown:
		return AxisVertical
	case DirectionLeft, DirectionRight:
		return AxisHorizontal
	default:
		return AxisVertical
	}
}

type Entity struct {
	Sprite *ebiten.Image

	Frame          float64
	FrameIncrement float64
	FrameTotal     float64
	FrameDirection map[Direction]int

	X          float64
	Y          float64
	Width      float64
	Height     float64
	Direction  Direction
	Directions []Direction
	Velocity   float64

	LastCorner *Entity

	BlueFrames int

	FlashFrames int
	Flash       bool
}

func (e *Entity) IsPower() bool {
	return e.Sprite == imageFood && e.Frame == 1
}

func NewEntityBee(tx, ty int, d Direction) *Entity {
	return &Entity{
		Sprite:         imageBee,
		FrameIncrement: 0.1,
		FrameTotal:     4,
		FrameDirection: map[Direction]int{
			DirectionDown:  0,
			DirectionUp:    1,
			DirectionLeft:  1,
			DirectionRight: 0,
		},
		X:         float64(8*tx + 4),
		Y:         float64(8*ty + 4),
		Width:     16,
		Height:    16,
		Direction: d,
		Velocity:  0.75,
	}
}

func EntityBear() *Entity {
	return &Entity{
		Sprite:         imageBear,
		FrameIncrement: 0.1,
		FrameTotal:     4,
		FrameDirection: map[Direction]int{
			DirectionDown:  0,
			DirectionUp:    1,
			DirectionLeft:  2,
			DirectionRight: 3,
		},
		X:         112,
		Y:         212,
		Width:     16,
		Height:    16,
		Direction: DirectionLeft,
		Velocity:  1.0,
	}
}

func EntityBees() []*Entity {
	return []*Entity{
		NewEntityBee(1, 4, DirectionRight),
		NewEntityBee(26, 4, DirectionLeft),
		NewEntityBee(1, 29, DirectionRight),
		NewEntityBee(26, 29, DirectionLeft),
	}
}
